ds_import/flver_extractor.py

```diff
@@ -0,0 +1,370 @@
+                    version_minor = reader.get_int8()
+                    reader.seek(10)
+                    if reader.get_int8() == 2 and version_minor == 12:      # is version 2.12?
+                        return True
+
+            return False
+
+
+    def __obtain_info(self, binary_reader, struct_fmt, item_count, named_tuple):
+        """
+        @type named_tuple: namedtuple
+        """
+        info_list = []
+        for x in range(0, item_count):
+            info_struct = binary_reader.get_struct(struct_fmt)
+            info_list.append(named_tuple._make(info_struct))
+        return tuple(info_list)
+
+
+    def __define_vertset_formats(self):
+
+        vertset_formats = []
+        with BinaryReader(self.__flver_file_path) as reader:
+            for vert_info in self.__vertex_description_info:
+                reader.seek(vert_info.description_offset)
+                formats = []
+                for i in range(0, vert_info.datatype_count):
+                    tempStruct = reader.get_struct("IIIII")
+                    typeConcat = str(tempStruct[2]) + str(tempStruct[3])
+
+                    if typeConcat == "20":
+                        formats.append("position")
+                    elif typeConcat == "172":
+                        formats.append("bone index")
+                    elif typeConcat == "193":
+                        formats.append("normal")
+                    elif typeConcat == "196":
+                        formats.append("vertex color")
+                    elif typeConcat == "1910":
+                        formats.append("bitangent")
+                    elif typeConcat == "225":
+                        formats.append("diffuse/lightmap UV")
+                    elif typeConcat == "261":
+                        formats.append("bone weight")
+                    elif typeConcat == "215":
+                        formats.append("diffuse UV")
+                    elif typeConcat == "197":
+                        formats.append("unknown")
+                vertset_formats.append(formats)
+        return vertset_formats
+
+    # sorts a given faceset into an order that Blender can use
+    def __sort_faces(self, faces):
+        # TODO: figure out why this works
+        faceslist = []
+        StartDirection = -1
+        f1 = faces[0]
+        f2 = faces[1]
+        FaceDirection = StartDirection
+        counter = 2
+        while counter < len(faces):
+
+            f3 = faces[counter]
+            FaceDirection *= -1
+            if (f1 != f2) and (f2 != f3) and (f3 != f1):
+                if FaceDirection > 0:
+                    faceslist.append(f1)
+                    faceslist.append(f2)
+                    faceslist.append(f3)
+                else:
+                    faceslist.append(f1)
+                    faceslist.append(f3)
+                    faceslist.append(f2)
+            counter = counter + 1
+            f1 = f2
+            f2 = f3
+
+        return faceslist
```

Comment:
This took me down a weird rabbit hole. I think because of the naming. This isn't sorting faces, it's sorting the _vertices_ of the faces, correct? Also the comment `# sorts a given faceset into an order that Blender can use` threw me off because I had already created a FaceSet class that does something completely different. 

What I think is happening with faces, and please correct me if I'm wrong, is that there's a long list of ints being read that are the indices of the vert_list. Every 3 vertex indices are then parsed into a Face each. The sort is needed to ensure vertex orders expected by blender. 

The algorithm appears to:
* for all vertex indices
  * reverse the order used for the last vertex
  * create a set of 3 vertex indices
  * ensure it does not contain duplicate vertex indices (if there is a duplicate, it is skipped)
  * append them to the list. 

The returned list is shorter than the original.

My suggested naming changes below. 


```suggestion
    # sorts a given list of vertex indices defining a set of faces into an order that Blender can use
    def __sort_faces_vertex_list(self, faces_vertex_list):

        # TODO: figure out why this works
        sorted_faces_vertex_list = []
        StartDirection = -1
        face_vertex1 = faces_vertex_list[0]
        face_vertex2 = faces_vertex_list[1]
        FaceDirection = StartDirection

        counter = 2
        while counter < len(faces_vertex_list):
            face_vertex3 = faces_vertex_list[counter]
            
            # For each vertex in the list, reverse the order of the vertices, 
            # effectively reversing the normal of the face.
            FaceDirection *= -1

            # If any of the three vertices in this set are duplicates, 
            # do not append, effectively skipping one vertex.
            if (face_vertex1 != face_vertex2) and (face_vertex2 != face_vertex3) and (face_vertex3 != face_vertex1):
                if FaceDirection > 0:
                    sorted_faces_vertex_list.append(face_vertex1)
                    sorted_faces_vertex_list.append(face_vertex2)
                    sorted_faces_vertex_list.append(face_vertex3)
                else:
                    sorted_faces_vertex_list.append(face_vertex1)
                    sorted_faces_vertex_list.append(face_vertex3)
                    sorted_faces_vertex_list.append(face_vertex2)
                    
            counter = counter + 1
            
            # Shift the set of vertices down the list by one
            face_vertex1 = face_vertex2
            face_vertex2 = face_vertex3

        return sorted_faces_vertex_list
```

Reply:
Yeah, you're right about "faces" in this case really being indices into the vertex list.

That looks like a solid breakdown for how this method is converting the native flver ordering into a blender-ready format.  What I'm very curious about is why it's stored that way in the Dark Souls data.  Seems wasteful, but there must be an explanation for it.  I may spend some time looking at it for fun later.

Reply:
"fun"